### ISSUE TITLE
Missing std::hash<enum> for map arguments

### ISSUE THREAD

**Issue by artwyman** [dropbox/djinni#213]
Details copied from discussion of issue #165.

Explanation of how to reproduce the problem from @fostah:

Here's how I edited the example project. To simplify things, I just added a new static method to sort_items:

```
sort_items = interface +c {
    static test(sortMap: map<sort_order, textbox_listener>);
    // ...
}
```

Then in sort_items_impl I just have a blank implementation of test:

```
void SortItems::test(const std::unordered_map<sort_order, std::shared_ptr<TextboxListener>> & sortMap) {

}
```

When I try to compile (make example_ios), I see the following error:

```
In file included from /Users/mfoster/Workspace/TEMP/dropbox/djinni/example/generated-src/cpp/sort_items.hpp:6:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:597:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/type_traits:964:38: error: implicit instantiation of undefined template 'std::__1::hash<textsort::sort_order>'
    : public integral_constant<bool, __is_empty(_Tp)> {};
```

Note that if I change the key to a string, everything compiles fine.

Explanation from @atwyman: This issue seems to exist when compiling using C++11 only.  Changing to c++1y in common.gypi makes the problem go away.  That plus the specifics of the error makes me realize that what's missing isn't the definition of the enum, but the definition of the specialized std::hash in the enum's header. It seems that in C++14 the STL already can hash enums by default, without help? I think that rings a bell as something I heard about in the past.

Regardless, I think the place where the definition is needed isn't necessarily in the sort_items.hpp header, but in TXSSortItems+Private.mm where the map actually needs to be created. It might make sense to generate it there, or in one of the +Private header files for ObjC++. I'll look into that.

**Comment by j4cbo** [comment]
The enum hash issue is LWG 2148 (http://www.open-std.org/jtc1/sc22/wg21/docs/lwg-defects.html#2148). It's fixed in the C++14 standard, and has been fixed in Clang for quite some time, but wasn't fixed in GCC until late 2015 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60970). In Djinni it's controlled by `cppEnumHashWorkaround`, see https://github.com/dropbox/djinni/blob/a656d150df84e21fe55f0eb64cd4dc99203b61a4/src/source/CppGenerator.scala#L74-L90

**Comment by fostah** [comment]
@j4cbo That makes sense why enums work as map keys when including the enum, instead of forward declaring it. Maybe generating an include for enums is the solution? I can't come up with a scenario where doing so would cause the circular inclusion that was addressed by #165.

Anyways, I appreciate the help on this issue.

**Comment by artwyman** [comment]
I'd rather avoid adding extra header includes in the hpp, on the general principle of minimizing dependencies to make compilation faster.  It should always be sufficient to include the enum in a cpp or mm file, and the generator can be tweaked to do that with a bit more effort.

**Comment by davidchappelle** [comment]
Any idea on an ETA for a fix?

**Comment by artwyman** [comment]
At this point, unfortunately, it's "as soon as I have spare time", and thus not likely to be too soon.  PRs are welcome, though. :)

If you need a workaround, C++14 will apparently work, or you could change the Scala code of Djinni to always generate full includes rather than forward-decls.  Just override the forwardDeclareOnly argument to true in CppGenerator.scala, and you should get something which will work, unless you're in the scenario where those includes caused circularity.

**Comment by steweg** [comment]
HI, one small add into this topic
even though in C++14 it should be fixed, not all implemented it correctly. We are already using C++14 version for our project and find out that if used with GCC compiler using standard on Debian packages (we currently use package 4.9.2-10 with gnustl++ heaeders), this issue is still present. In order to create a workaround, we create a special header file that is included only if compiler is GCC, which practically moved what was already in LLVM/clang headers (libc++) into our code, Something like [hash-enum.hpp.txt](https://github.com/dropbox/djinni/files/317847/hash-enum.hpp.txt)

So if your projects depends on this kind of staff, feel free to use similar workaround

**Comment by artwyman** [comment]
I've lost some context on this, but I believe this was fixed by #259 and am thus closing it out.  Please re-open if you think there's more to be fixed.
